src/main/scala/scorex/core/validation/ValidationResult.scala

```diff
@@ -0,0 +1,68 @@
+package scorex.core.validation
+
+import akka.http.scaladsl.server.Route
+import scorex.core.api.http.ApiError
+import scorex.core.validation.ValidationResult.{Invalid, Valid}
+
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+/** Base trait for the result of validation
+  */
+sealed trait ValidationResult {
+
+  def isValid: Boolean
+
+  def message: String
+
+  def errors: Seq[ModifierError]
+
+  def ++(next: ValidationResult): ValidationResult
+
+  def toTry: Try[Unit]
+
+  def toFuture: Future[Unit] = Future.fromTry(toTry)
+
+  def toApi(onSuccess: => Route): Route = this match {
+    case Valid => onSuccess
+    case Invalid(_) => ApiError.BadRequest(message)
+  }
+
+}
+
+object ValidationResult {
+
+  type Valid = ValidationResult.Valid.type
+
+  /** Successful validation result
+    */
+  final case object Valid extends ValidationResult {
+    def isValid: Boolean = true
+    def message: String = "OK"
+    def errors: Seq[ModifierError] = Seq.empty
+    def ++(next: ValidationResult): ValidationResult = next
+    def toTry: Try[Unit] = Success(())
+  }
+
+  /** Unsuccessful validation result
+    */
+  final case class Invalid(errors: Seq[ModifierError]) extends ValidationResult {
+    def isValid: Boolean = false
+    def isFatal: Boolean = errors.exists(_.isFatal)
+    def message: String = "Validation errors: " + errors.mkString(" | ")
+
+    def ++(next: ValidationResult): ValidationResult = {
+      next match {
+        case Valid => this
+        case Invalid(e2) => Invalid(errors ++ e2)
+      }
+    }
+
+    @SuppressWarnings(Array("org.wartremover.warts.TraversableOps"))
+    lazy val toTry: Try[Unit] = {
+      if (errors.size == 1) Failure(errors.head.toThrowable) else Failure(MultipleErrors(errors))
+    }
```

Comment:
In this definition I wonder how to report what I guess is an inconsistent state where `errors` is empty.
Also, do we need to separate the case of a single error and multiple errors? Can't we rename and use `MultipleErrors` for both scenarios?

Reply:
we could use `NonEmptyList` from cats in `MultipleErrors` but I don't really think that we need this complication. If somebody accidentally creates MultipleErrors with empty list, well this will not have any fatal consequences.

I'm not wrapping single error to `MultipleErrors` class to make the stacktrace more clean. It's easier to see the cause of an exception immediately in log, not looking for nested exception.

As for making all error Multiple i don't thing it's a good idea, because in this case we could have something looking like tree of exceptions inside, not a list

Reply:
I agree that using `NonEmptyList` would be too much for this. I initially focused on this snippet when I saw the `SuppressWarnings` annotation. I think that it will be better to keep the logic simple as you stated rather than focus too much on the empty list detail.
Thank you for the time to reply to my comments.

Reply:
It looks like we are reinventing the wheel here. Couldn't we use cats `Validated` (https://typelevel.org/cats/datatypes/validated.html) to accumulate errors?

Reply:
I've started this piece from trying to apply Cats Validated for modifier validation in Ergo, but it didn't suit. It's supposed to be used mostly for the web from validation. It's really good for accumulating all the validated fields and after that constructing a composite object from all these fields. 

We have a pretty different case, because we need to perform multiple checks for the same object without any transformation. This looks too messy with cats. See the example of that kind of validation in Ergo HeaderValidator here: https://github.com/ergoplatform/ergo/blob/i269_post_transatction/src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala#L300.

Also cats `Validated` doesn't support fail-fast validation, but we do.

Reply:
Ok. Thanks for explanation, @mike-aksarin !

I would suggest adding your explanation as a comment in the code. Otherwise other people will wonder why cats's `Validated` wasn't used.

Reply:
Done!